src/linear_algebra/finite_dimensional.lean

```diff
@@ -54,44 +94,94 @@ begin
     rw [set.finite.coe_to_finset, ← hb.2], refl }
 end
 
+/-- The dimension of a finite-dimensional vector space, as a cardinal, is strictly less than the
+first infinite cardinal `omega`. -/
 lemma dim_lt_omega (K V : Type*) [discrete_field K] [add_comm_group V] [vector_space K V] :
   ∀ [finite_dimensional K V], dim K V < omega.{v} :=
 finite_dimensional_iff_dim_lt_omega.1
 
-set_option pp.universes true
-
-lemma of_fg (hfg : (⊤ : submodule K V).fg) : finite_dimensional K V :=
-let ⟨s, hs⟩ := hfg in
-begin
-  rw [finite_dimensional_iff_dim_lt_omega, ← dim_top, ← hs],
-  exact lt_of_le_of_lt (dim_span_le _) (lt_omega_iff_finite.2 (set.finite_mem_finset s))
-end
-
-lemma exists_is_basis_finite (K V : Type*) [discrete_field K]
-  [add_comm_group V] [vector_space K V] [finite_dimensional K V] :
+/-- In a finite dimensional space, there exists a finite basis. A basis is in general given as a
+function from an arbitrary type to the vector space. Here, we think of a basis as a set (instead of
+a function), and use as parametrizing type this set (and as a function the function `subtype.val`).
+-/
+variables (K V)
+lemma exists_is_basis_finite [finite_dimensional K V] :
   ∃ s : set V, (is_basis K (subtype.val : s → V)) ∧ s.finite :=
 begin
   cases exists_is_basis K V with s hs,
   exact ⟨s, hs, finite_of_linear_independent hs.1⟩
 end
+variables {K V}
 
-instance [finite_dimensional K V] (S : submodule K V) : finite_dimensional K S :=
+/-- A vector space if finite-dimensional if and only if it is finitely generated. As the
+finitely-generated property is a property of submodules, we formulate this in terms of the
+maximal submodule, equal to the whole space as a set by definition.-/
+lemma iff_fg :
+  finite_dimensional K V ↔ (⊤ : submodule K V).fg :=
+begin
+  split,
+  { assume h,
+    haveI : finite_dimensional K V := h,
```

Comment:
If you want, you can change this into `resetI`, and I would even find it legitimate to tack it onto the same line as `assume h`. But maybe you prefer it this way for readability? We should have an `introI` tactic.

Reply:
There is an `introI` tactic!